internal/controller/openstackcloud_controller.go

```diff
@@ -0,0 +1,272 @@
+type OpenStackCloudReconciler struct {
+	client.Client
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+func finalizerName(cloud *openstackv1.OpenStackCloud) string {
+	return OpenStackCloudFinalizer + "/" + cloud.Name
+}
+
+//+kubebuilder:rbac:groups=openstack.gopherkube.dev,resources=openstackclouds,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=openstack.gopherkube.dev,resources=openstackclouds/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=openstack.gopherkube.dev,resources=openstackclouds/finalizers,verbs=update
+//+kubebuilder:rbac:resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:resources=secrets/finalizers,verbs=update
+
+func (r *OpenStackCloudReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	logger := log.FromContext(ctx)
+	logger = logger.WithValues("OpenStackCloud", req.Name)
+
+	openStackResource := &openstackv1.OpenStackCloud{}
+	err := r.Client.Get(ctx, req.NamespacedName, openStackResource)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("resource not found in the API")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("reconciling resource")
+
+	patchResource := &openstackv1.OpenStackCloud{}
+	patchResource.TypeMeta = openStackResource.TypeMeta
+	util.InitialiseRequiredConditions(patchResource)
+	controllerutil.AddFinalizer(patchResource, OpenStackCloudFinalizer)
+
+	defer func() {
+		// If we're returning an error, report it as a TransientError in the Ready condition
+		if reterr != nil {
+			updated, condition := util.SetNotReadyConditionTransientError(openStackResource, patchResource, reterr.Error())
+
+			// Emit an event if we're setting the condition for the first time
+			if updated {
+				util.EmitEventForCondition(r.Recorder, openStackResource, corev1.EventTypeWarning, condition)
+			}
+		}
+
+		primaryPatch := &openstackv1.OpenStackCloud{}
+		primaryPatch.TypeMeta = patchResource.TypeMeta
+		primaryPatch.Labels = patchResource.Labels
+		primaryPatch.Finalizers = patchResource.Finalizers
+		// We must exclude the spec from the patch as it
+		// contains required fields which must not be included
+		applyerr := util.Apply(ctx, r.Client, openStackResource, primaryPatch, "spec")
+
+		if applyerr == nil {
+			statusPatch := &openstackv1.OpenStackCloud{}
+			statusPatch.TypeMeta = openStackResource.TypeMeta
+			statusPatch.Status = patchResource.Status
+			applyerr = util.ApplyStatus(ctx, r.Client, openStackResource, statusPatch)
+
+			// Ignore the error if we get NotFound after removing the finalizer
+			if applyerr != nil && apierrors.IsNotFound(applyerr) && len(primaryPatch.Finalizers) == 0 {
+				applyerr = nil
+			}
+		}
+
+		reterr = errors.Join(reterr, applyerr)
+	}()
+
+	if !openStackResource.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, logger, openStackResource, patchResource)
```

Comment:
nit: do you mind re-wrapping the logger with the context and we-extracting it into `reconcileDelete` like we do in the [other controllers](https://github.com/gophercloud/gopherkube/blob/3df81034156b271c023f3d7b523d71a8c4bea715/internal/controller/openstacksecuritygrouprule_controller.go#L108)?